ush/ocnpost.py

```diff
@@ -0,0 +1,646 @@
+#     not identical between the two grids.
+# 
+# c) conservative regridding of some fields
+#    Fields such as ice concentration or fluxes which inherently
+#    area area-weighted require conservative regridding. Most other
+#    variables are state variables and can be regridded using
+#    bilinear weighting.
+#
+# An efficient way to accomplish the unstaggering of velocities 
+# is to use the bilinear interpolation weights between grid 
+# points of the Arakawa C grid and the center grid points (for example 
+# Cu->Ct). These weights are generated by the weight generation script 
+#
+# Remapping from the tripole to rectilinear uses either the bilinear
+# or conservative weights from the weight generation script.  Bilinear weights 
+# generated for the first vertical level can be used on other levels
+# (where the masking changes) by utilizing the correct masking procedure.
+# Set output_masks to true to examine the interpolation masks.
+# 
+# Intermediate file output can easily be generated for debugging by 
+# follwing the example in the output_masks logical
+#
+# Bin.Li@noaa.gov (May 2019)
+# The scripts is revised for use in the coupled workflow.
+#
+# Eric.Sinsky@noaa.gov (June 2023)
+# The script has been ported from NCL to Python. 
+#----------------------------------------------------------------------
+
+from time import process_time
+import os
+import sys
+import numpy as np
+import netCDF4 as nc
+import numpy.f2py
+sys.path.append(os.getcwd()) #add working directory
+
+def initfortranprog():
+    fsource1 = '''
+        subroutine bilin(src,S,row,col,frac_b,dst_len,dst,src_len,n_s)
+           implicit none
+           integer, intent(in) :: dst_len
+           integer, intent(in) :: row(n_s)
+           integer, intent(in) :: col(n_s)
+           integer :: i,src_len, n_s
+           real(kind=8), intent(in) :: frac_b(dst_len)
+           real(kind=8), intent(in) :: src(src_len)
+           real(kind=8), intent(in) :: S(n_s)
+           real(kind=8), intent(out) :: dst(dst_len)
+
+           !print*,"Starting bilinear interpolation"
+           !print*,'n_s =',n_s
+           !print*,'src_len =',src_len
+           !print*,'dst_len =',dst_len
+
+           ! Initialize destination field to 0.0
+           do i=1,dst_len
+             dst(i)=0.0
+           enddo
+
+           ! Apply weights
+           do i=1, n_s
+             dst(row(i))=dst(row(i))+S(i)*src(col(i))
+           enddo
+           
+           do i=1, dst_len
+             if (frac_b(i) .ne. 0.0) then
+               dst(i)=dst(i)/frac_b(i)
+             endif
+           enddo
+
+        end subroutine
+    '''
+
+    try:
+        import esmfwghtinterp_f90subroutine
+    except ImportError or ModuleNotFoundError:
+        print('Weighted interpolation fortran subroutine has not been compiled. Compiling...')
+        np.f2py.compile(fsource1, modulename='esmfwghtinterp_f90subroutine', verbose=False)
+        import esmfwghtinterp_f90subroutine
```

Comment:
One cannot be compiling everytime this script is run.

Reply:
I agree. As of now, a compilation is only performed if the .so file does not exist in any Python path. However, the compile step can be removed completely from the code if that is preferred.

Reply:
I tried something very similar to this. The program ran less efficiently when this was a python function.

Reply:
There are efficient bilinear interpolation methods in `numpy`.  Did you try any of those? 
If this is the meat and potatoes of this program, IMO, it makes sense for this to be a Fortran program.  The rest is just IO.

Reply:
I tried [esmpy.Regrid](https://earthsystemmodeling.org/esmpy_doc/release/ESMF_8_0_1/html/regrid.html), but this also turned out to be less efficient.

Reply:
I meant something from `numpy` or `scipy` e.g. https://numpy.org/devdocs/reference/generated/numpy.interp.html

Reply:
To my understanding, `np.interp` does not have an option for a weighted bilinear interpolation. This Fortran subroutine, which uses the ESMF regridding weights in a bilinear or conservative interpolation is referenced here [here](https://earthsystemmodeling.org/docs/release/ESMF_8_0_1/ESMF_refdoc/node3.html#SECTION03029300000000000000).

Reply:
I would personally avoid `numpy` and/or `scipy`. I have also tried these methods and they are slow and do not take into account issues such as landmask and fractional grids.

To reiterate, once you have the ESMF weights, all you need to do is add a short subroutine to read in the source grid variable and remap it to the destination grid. It should be short (on the order of 10 lines).

Reply:
@HenryWinterbottom-NOAA Thank you for your feedback. Just to confirm, is this the short subroutine that you are referring to? 

           ! Initialize destination field to 0.0
           do i=1,dst_len
             dst(i)=0.0
           enddo

           ! Apply weights
           do i=1, n_s
             dst(row(i))=dst(row(i))+S(i)*src(col(i))
           enddo
           
           do i=1, dst_len
             if (frac_b(i) .ne. 0.0) then
               dst(i)=dst(i)/frac_b(i)
             endif
           enddo

Reply:
@EricSinsky-NOAA Yes.